### PR TITLE
apps/examples/timer: dedicated timer thread with sigwait()

### PR DESCRIPTION
**Pull request by NitishAmbastha**
This patch is to mitigate nested signal issues with
high-resolution timer interrupts. Nested signal situation
occurs if the time-period is smaller and next timer interrupt
occurs before the current interrupt signal is delivered.
This causes stack corruption due to multiple pending signals

This patch implements a dedicated thread that waits on sigwait()
It provides a deterministic wake-up time when the signal occurs
With this, we can achieve upto 30us timer events without suppressing
timer interrupts.

For more resolution requirement, we need to increase the stack for the thread waiting for signal

### CODE REVIEW

**Comment by seinfra** [comment]
Target : [c2d3e1414c15aa3c5aed193fcdae31ac747fc3e0] - Code Rule Check Result:
apps/examples/timer/timer_main.c:104: ERROR: [SPC_M_OPR] space required after that ',' (ctx:VxV)
apps/examples/timer/timer_main.c:106: ERROR: [SPC_M_OPR] space required after that ',' (ctx:VxO)
apps/examples/timer/timer_main.c:106: ERROR: [SPC_M_OPR] space required before that '&' (ctx:OxV)
apps/examples/timer/timer_main.c:106: ERROR: [SPC_M_OPR] space required after that ',' (ctx:VxV)
apps/examples/timer/timer_main.c:108: ERROR: [SPC_M_KWD] space required before the open parenthesis '('
apps/examples/timer/timer_main.c:114: ERROR: [SPC_M_OPR] space required after that ',' (ctx:VxO)
apps/examples/timer/timer_main.c:114: ERROR: [SPC_M_OPR] space required before that '&' (ctx:OxV)
apps/examples/timer/timer_main.c:114: ERROR: [SPC_M_OPR] space required after that ',' (ctx:VxV)

**Comment by seinfra** [comment]
Target : [c2d3e1414c15aa3c5aed193fcdae31ac747fc3e0] - Code Rule Check (C++) OK.

**Comment by jeongchanKim** [comment]
Could you change the branch to master, not protection?

**Comment by seinfra** [comment]
Target : [d97e711e1243a5172ff1a205313243c8de8f37c4] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [d97e711e1243a5172ff1a205313243c8de8f37c4] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [d9553b3bb9de1136088e7ca8f9f7ad6d904d2ce5] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [d9553b3bb9de1136088e7ca8f9f7ad6d904d2ce5] - Code Rule Check OK.

File: apps/examples/timer/timer_main.c [hunk 1]
```diff
@@ -218,7 +201,7 @@ int timer_main(int argc, char *argv[])
 	while ((opt = getopt(argc, argv, "t:n:f:h")) != -1) {
 		switch (opt) {
 		case 't':
-			intval = strtol(optarg, NULL, 10) * MSEC_PER_SEC;
+			intval = strtol(optarg, NULL, 10);
```
Comment (amandeepc) [inline apps/examples/timer/timer_main.c]: Please update this info in USAGE message as well

File: apps/examples/timer/timer_main.c [hunk 2]
```diff
@@ -90,57 +90,40 @@
...
-	 * NOTE: The use of signal handler is not recommended if you are concerned
-	 * about the signal latency.  Instead, a dedicated, high-priority thread
-	 * that waits on sigwaitinfo() is recommended.  High priority is required
-	 * if you want a deterministic wake-up time when the signal occurs.
-	 */
-	sem_post(&g_sem);
-}
 
-static pthread_addr_t timer_thread(pthread_addr_t arg)
+static pthread_addr_t hrt_thread(pthread_addr_t arg)
```
[the remark this answers is not in the document]
Comment (sunghan-chang) [reply apps/examples/timer/timer_main.c]: How about guiding both?
sigwait for hrt
sigaction for lrt

If you think it looks not good, let's rename this function.

File: apps/examples/timer/timer_main.c [hunk 3]
```diff
@@ -253,8 +236,12 @@ int timer_main(int argc, char *argv[])
 		repeat = EXAMPLE_TIMER_NSAMPLES;
 	}
 
-	/* Create Time Handler Thread */
-	tid = create_timer_thread((void *)repeat);
+	sigemptyset(&sigset);
+	sigaddset(&sigset, EXAMPLE_TIMER_SIGNO);
+	sigprocmask(SIG_BLOCK, &sigset, NULL);
```
Comment (sunghan-chang) [inline apps/examples/timer/timer_main.c]: Why is this necessary here?

File: apps/examples/timer/timer_main.c [hunk 4]
```diff
@@ -90,57 +90,40 @@
...
+	}
+	fprintf(stdout, "signal received:%d\n", (int)arg);
+	pthread_sigmask(SIG_UNBLOCK, &sig_set, NULL);
 	pthread_exit(NULL);
 
 	return NULL;
 }
 
-static pthread_t create_timer_thread(void *arg)
+static pthread_t create_hrt_thread(void *arg)
```
Comment (sijupunnnoose) [inline apps/examples/timer/timer_main.c]: Keep same earlier name, i dont think we should use hrt here.

File: apps/examples/timer/timer_main.c [hunk 5]
```diff
@@ -166,13 +149,13 @@ static pthread_t create_timer_thread(void *arg)
 		return -ret;
 	}
 
-	ret = pthread_create(&tid, &attr, timer_thread, arg);
+	ret = pthread_create(&tid, &attr, hrt_thread, arg);
```
Comment (sijupunnnoose) [inline apps/examples/timer/timer_main.c]: ditto

File: apps/examples/timer/timer_main.c [hunk 6]
```diff
@@ -253,8 +236,12 @@ int timer_main(int argc, char *argv[])
...
 	}
 
-	/* Create Time Handler Thread */
-	tid = create_timer_thread((void *)repeat);
+	sigemptyset(&sigset);
+	sigaddset(&sigset, EXAMPLE_TIMER_SIGNO);
+	sigprocmask(SIG_BLOCK, &sigset, NULL);
+
+	/* Create HRT Thread */
+	tid = create_hrt_thread((void *)repeat);
```
Comment (sijupunnnoose) [inline apps/examples/timer/timer_main.c]: ditto

File: apps/examples/timer/timer_main.c [hunk 7]
```diff
@@ -90,57 +90,40 @@
...
-		fprintf(stdout, "time limits(%d)\n", ++nbr);
-	}
+	sigset_t sig_set;
+	sigemptyset(&sig_set);
+	sigaddset(&sig_set, EXAMPLE_TIMER_SIGNO);
+
+	pthread_sigmask(SIG_BLOCK, &sig_set, NULL);
 
+	while (count--) {
+		fprintf(stdout, "Waiting on a signal...\n");
```
Comment (sijupunnnoose) [inline apps/examples/timer/timer_main.c]: Avoid so many prints, just keep as earlier.

**Comment by seinfra** [comment]
Target : [4d415d09797d9e95f4771834ed79526432aaa086] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [4d415d09797d9e95f4771834ed79526432aaa086] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [e05c83c08f5381ff89ac1374e798a495e2ddf675] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [e05c83c08f5381ff89ac1374e798a495e2ddf675] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [d65da425a2ad317cae59e15a5b1d847be0feb648] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [d65da425a2ad317cae59e15a5b1d847be0feb648] - Code Rule Check OK.

File: apps/examples/timer/timer_main.c (hunk 1 above)
Comment (NitishAmbastha) [reply apps/examples/timer/timer_main.c]: I would like to keep the input in ms. So, I reverted this part of the change

File: apps/examples/timer/timer_main.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (NitishAmbastha) [reply apps/examples/timer/timer_main.c]: Since this implementation should work well with low resolution as well as high resolution, I renamed the function as generic name

File: apps/examples/timer/timer_main.c (hunk 3 above)
Comment (NitishAmbastha) [reply apps/examples/timer/timer_main.c]: You are right. This is not necessary, we are explicitly creating the signal set in the timer thread. Cleanup done.

File: apps/examples/timer/timer_main.c (hunk 4 above)
Comment (NitishAmbastha) [reply apps/examples/timer/timer_main.c]: Done

File: apps/examples/timer/timer_main.c (hunk 5 above)
Comment (NitishAmbastha) [reply apps/examples/timer/timer_main.c]: Done

File: apps/examples/timer/timer_main.c (hunk 6 above)
Comment (NitishAmbastha) [reply apps/examples/timer/timer_main.c]: Done

File: apps/examples/timer/timer_main.c (hunk 7 above)
Comment (NitishAmbastha) [reply apps/examples/timer/timer_main.c]: Done

**Comment by NitishAmbastha** [comment]
Updated the title template

**Comment by sunghan-chang** [comment]
@NitishAmbastha Could you make getting time interval in micro second as you did? In milisecond, this change is unnecessary.

File: apps/examples/timer/timer_main.c (hunk 1 above)
Comment (sunghan-chang) [reply apps/examples/timer/timer_main.c]: I don't agree reverting. The main purpose of this PR is supporting in microsecond. But if you make it working in milisecond, no need to change.

**Comment by seinfra** [comment]
Target : [0c072cc048478311517f1064d38ee68490ec38d1] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [0c072cc048478311517f1064d38ee68490ec38d1] - Code Rule Check OK.

**Comment by NitishAmbastha** [comment]
> @NitishAmbastha Could you make getting time interval in micro second as you did? In milisecond, this change is unnecessary.

Done
